File: WuV/Exercises/arith_semantics.java
/*
 one abstract class per non-terminal

   for every non-terminal
     one extending class per production

   for every production
     one constructor argument per non-terminal on rhs
*/

abstract class N {};

class one extends N {
    one() {};
}
class zero extends N {
    zero() {};
}
class plus extends N {
    public N left;
    public N right;
    plus(N left, N right) {
        this.left = left;
        this.right = right;
    }
}

/* in Scala, the Java boilerplate disappears and the principle is clearer
   class plus(left: N, right: N) extends N
*/

class times extends N {
    public N left;
    public N right;
    times(N left, N right) {
        this.left = left;
        this.right = right;
    }
}

abstract class F {};

class equal extends F {
    public N left;
    public N right;
    equal(N left, N right) {
        this.left = left;
        this.right = right;
    }
}
class leq extends F {
    public N left;
    public N right;
    leq(N left, N right) {
        this.left = left;
        this.right = right;
    }
}

class main {
    public static void main(String[] args) {
        N example = new times(
            new plus(new one(), new one()),
            new plus(new one(), new plus(new one(), new one()))
        );
        System.out.println(Semantics.semN(example));
    }
}

/* Semantics:
    as for SML
*/

class Semantics {
   public static String semN(N x) {
       if (x instanceof zero) {
           return "";
       } else if (x instanceof one) {
           return "|";
       } else if (x instanceof plus) {
           plus p = (plus) x;
           return semN(p.left) + semN(p.right);
       } else if (x instanceof times) {
           times p = (times) x;
           return semN(p.left).replace("|", semN(p.right));
       } else {
           // forbidden case, raise error
           return null;
       }
   }

   public static Boolean semF(F x) {
       if (x instanceof equal) {
           equal p = (equal) x;
           return semN(p.left) == semN(p.right);
       } else if (x instanceof leq) {
           leq p = (leq) x;
           return semN(p.left).startsWith(semN(p.right));
       } else {
           // forbidden case, raise error
           return false;
       }
   }
}

File: WuV/Exercises/arith_semantics.ml
(* Syntax:

  one inductive data type per non-terminal

  for every non-terminal:
    one constructor per production

  for every production:
    one constructor argument for non-terminal on the rhs of the production

*)

datatype N = zero
           | one
           | plus of N*N
           | times of N*N

     and F = equal of N*N
           | leq of N*N

val example: N =
  times(
    plus(one,one),
    plus(one,plus(one,one))
  )
;

times(example,example);
plus(it,one);

(* Semantics

  one induction function for every non-terminal
  
  for every non-terminal:
    one case per production

    for every production:
      one recursive call per non-terminal on the rhs
*)

fun semN(zero) = ""
  | semN(one) = "|"
  | semN(plus(m,n)) = semN(m) ^ semN(n)
  | semN(times(m,n)) = String.translate(fn x => semN n) (semN m)

;

fun semF(equal(m,n)) = (semN m) = (semN n)
  | semF(leq(m,n)) = String.isPrefix (semN m) (semN n)
;

semN(example);
semF(leq(one,example));

(* alternative function definitions without syntactic sugar *)

fun semNA(x:N): string =
 case x of
    zero => ""
  | one => "|"
  | plus(m,n) => (semNA m) ^ (semNA n)
  | times(m,n) => String.translate(fn x => semNA n) (semNA m)

fun semFA(x: F): bool =
 case x of
    equal(m,n) => (semNA m) = (semNA n)
  | leq(m,n) => String.isPrefix (semNA m) (semNA n)
;

semNA(example);
semFA(leq(one,example));
